src/luaotfload-auxiliary.lua

```diff
@@ -105,6 +105,14 @@ luaotfload_callbacks [#luaotfload_callbacks + 1] = {
   "patch_font", set_sscale_dimens, "set_sscale_dimens",
 }
 
+-- HarfteX doesn't like 1-based indexing, so we adjust it. This doesn't hurt in
+-- LuaTeX either because LuaTeX doesn't use a `index` field in the first place.
+luaotfload_callbacks [#luaotfload_callbacks + 1] = {
+  "patch_font", function(fontdata)
+    fontdata.index = (fontdata.specification.sub or 1) - 1
```

Comment:
I can make the Lua API 1-based if this looks better, but the C code is 0-based.

Reply:
@khaledhosny Personally I think 1-based would look more natural in Lua. (Subfonts are kind of like an array of fonts, and arrays are 1-based in Lua) But as long as LuaTeX doesn't start using the same field as a 1-based index, I don't really care.

Reply:
The font `characters` table seems to be 0-based, and that one have stronger reasons to be 1-based, so I think I'm sticking to 0 for now.

Reply:
Actually the `characters` table doesn't behave like an array, especially the indices are not continuous. It is a associative directory indexed by Unicode Codepoints and Unicode codepoints. (or some other encoding) These Encodings are 0-based, so I actually can't think of any reason why characters should be 0-based. Anyway, harftex has harf and not lua in it's name, so conforming more to HarfBuzz than to LuaTeX conventions probably makes sense.